CDS-94177 (#5965)
harness/developer-hub: release-notes/continuous-delivery.md

## March 2024

### Version 1.30.x

#### Early access

> - You can create a multi-phase workflow that progressively deploy your new instances to a new ASG incrementally using the ASG Phased Deploy step when creating a Canary deployment. Currently, this feature is behind the feature flag, `CDS_ASG_PHASED_DEPLOY_FEATURE_NG`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. For more information, go to [Canary phased deployment](/docs/continuous-delivery/deploy-srv-diff-platforms/aws/asg-tutorial/#canary-phased-deployment). (CDS-87684)

Reviewer: ```suggestion
- You can create a multi-phase workflow that progressively deploys your new instances to a new ASG incrementally using the ASG Phased Deploy step when creating a Canary deployment. Currently, this feature is behind the feature flag, `CDS_ASG_PHASED_DEPLOY_FEATURE_NG`. Contact [Harness Support](mailto:support@harness.io) to enable the feature. For more information, go to [Canary phased deployment](/docs/continuous-delivery/deploy-srv-diff-platforms/aws/asg-tutorial/#canary-phased-deployment). (CDS-87684)
```

- The wrapper for matrix nodes appeared successful even if the matrix nodes were skipped. This issue is resolved by adding a check to see the status of a strategy node's children. If all child nodes are skipped, the strategy node will be skipped now. (CDS-92727)
- Marking failure strategy for a step with in a stage as ignore failure did not work. The stage status was not shown as success of the step in the stage failed. This issue is fixed. Now, `IgnoreFailed` status is considered as Success for sending the stage notifications. (CDS-92057, ZD-58259)
- Stage and pipeline are not marked as ResourceWaiting even if the stage is waiting on the resources. This issue is fixed by creating a `ResourceWaitingStepStatusUpdateHandler` logic that will mark the stage and pipeline as `ResourceWaiting`. When the step resumes, the existing `handleStepResponseInternal` logic marks the stage and pipeline as running. (CDS-87769)

> - If the `serverless print` command used to resolve all Serverless variables and get the stack name is not configured inside the Serverless YAML, Harness used the default stack name, `<serviceNameInYaml> - <stageNameSpecifiedInHarnessInfra>`. So, even when the `serverless print` command failed, the step did not fail, and Harness considered the default stack name, which is wrong. This issue is fixed by updating the Serverless Prepare Rollback step image to version, `harnessdev/serverless-preparerollback:3.30.1-2.0.0`. (CDS-87684)

Reviewer 2: In serverless prepare rollback  image, we would use "serverless print" command . This command is used to resolve all the Serverless variables and fetch the custom stack name if provided in serverless yaml. Earlier, even if the serverless print command failed, we would not fail the step , and consider the stack name as `<serviceNameInYaml> - <stageNameSpecifiedInHarnessInfra>'.  But with this new image, if serverless print command fails, we would fail the step.

> - If the `serverless print` command used to resolve all Serverless variables and get the stack name is not configured inside the Serverless YAML, Harness used the default stack name, `<serviceNameInYaml> - <stageNameSpecifiedInHarnessInfra>`. So, even when the `serverless print` command failed, the step did not fail, and Harness considered the default stack name, which is wrong. This issue is fixed by updating the Serverless Prepare Rollback step image to version, `harnessdev/serverless-preparerollback:3.30.1-2.0.0`. (CDS-87684)

Author: ```suggestion
- In the Serverless Prepare Rollback image, the `serverless print` command is used to resolve all Serverless variables and fetch custom stack name, if provided in the Serverless YAML. Earlier, even if the `serverless print` command failed, Harness did not fail the step, and considered the stack name as `-`. This issue is fixed by updating the Serverless Prepare Rollback step image to version, `harnessdev/serverless-preparerollback:3.30.1-2.0.0`. With this new image, Harness will fail the step if the `serverless print` command failed. (CDS-87684)
```

> - In the Serverless Prepare Rollback image, the `serverless print` command is used to resolve all Serverless variables and fetch custom stack name, if provided in the Serverless YAML. Earlier, even if the `serverless print` command failed, Harness did not fail the step, and considered the stack name as `-`. This issue is fixed by updating the Serverless Prepare Rollback step image to version, `harnessdev/serverless-preparerollback:3.30.1-2.0.0`. With this new image, Harness will fail the step if the `serverless print` command failed. (CDS-87684)

Author: ```suggestion
- In the Serverless Prepare Rollback image, the `serverless print` command is used to resolve all Serverless variables and fetch custom stack name, if provided in the Serverless YAML. Earlier, even if the `serverless print` command failed, Harness did not fail the step, and considered the stack name as `<serviceNameInYaml> - <stageNameSpecifiedInHarnessInfra>`. This issue is fixed by updating the Serverless Prepare Rollback step image version to`harnessdev/serverless-preparerollback:3.30.1-2.0.0`. With this new image, Harness will fail the step if the `serverless print` command failed. (CDS-87684)
```

  
  The following changes are implemented as part of this new image: 

>   1. The `serverless print` command should not fail due to plugins being specified in the Serverless YAML. With this new image, Harness does an NPM install (on package.json) before running the `serverless print` command. Make sure that the plugins are specified in the package.json to install them before running the `serverless print` command. Now, if your plugins are specified in the package.json at the same level where your Serverless YAML lies, your plugins would be installed, and correct stack name from Serverless YAML would be fetched. 

Author: ```suggestion
  1. The `serverless print` command would not fail due to plugins being specified in the Serverless YAML. With this new image, Harness does an NPM install (on package.json) before running the `serverless print` command. Make sure that the plugins are specified in the package.json to install them before running the `serverless print` command. Now, if your plugins are specified in the package.json at the same level where your Serverless YAML lies, your plugins would be installed, and correct stack name from Serverless YAML would be fetched. 
```

>   2. If the `serverless print` command still failed, Harness will fail the step and will not assume the default stack name.

Reviewer 2: "fails" instead of "failed"

Reviewer 2: Also specify the default stack name would be <serviceNameInYaml> - <stageNameSpecifiedInHarnessInfra>

>   2. If the `serverless print` command still failed, Harness will fail the step and will not assume the default stack name.

Author: ```suggestion
  2. If the `serverless print` command still fails, Harness will fail the step and will not assume the default stack name, `<serviceNameInYaml> - <stageNameSpecifiedInHarnessInfra>`.
```
